### ISSUE TITLE
typeof(GenericType<>) can produce invalid IL and cause BadImageFormatException at runtime

### ISSUE THREAD

**Issue by LeroyK** [dotnet/roslyn#38344]
Sometimes using ``typeof(GenericType<>)`` produces invalid IL. This will result in a BadImageFormatException at run time as soon as the method is called. The exception is probably thrown by RyuJit, because it can't process invalid IL.

I'm not sure what the conditions are, but you can find the repro below. When you remove ``Console.WriteLine("Hello World!");`` it produces correct IL.

**Exception**
```
Message: An attempt was made to load a program with an incorrect format. (0x8007000B)
Stack Trace:
   at RoslynBug.Program.Test(Type type) in RoslynBug\RoslynBug\Program.cs:line 19
   at RoslynBug.Program.Main(String[] args) in RoslynBug\RoslynBug\Program.cs:line 11
```

**Version Used**: 
dotnet-sdk-3.0.100-preview9-19414-10

**Steps to Reproduce**:
```csproj
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <LangVersion>preview</LangVersion>
    <Nullable>enable</Nullable>
    <Deterministic>true</Deterministic>
    <PathMap>$(SolutionDir)=$(SolutionName)</PathMap>
    <Features>pdb-path-determinism</Features>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <RuntimeIdentifier>win10-x64</RuntimeIdentifier>
    <TargetLatestRuntimePatch>true</TargetLatestRuntimePatch>
    <TieredCompilation>true</TieredCompilation>
    <TieredCompilationQuickJit>true</TieredCompilationQuickJit>
    <LinkDuringPublish>false</LinkDuringPublish>
    <CrossGenDuringPublish>true</CrossGenDuringPublish>
    <PublishReadyToRun>false</PublishReadyToRun>
  </PropertyGroup>
</Project>

```

```c#
using System;
using System.Collections.Generic;

namespace RoslynBug
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            Test(typeof(IDictionary<string, long>));
        }

        static Type[] Test(Type type)
        {
            var genericArguments = type.GetGenericArguments();
            genericArguments[0] = typeof(Example<>).MakeGenericType(genericArguments[0]);
            return genericArguments;
        }

        static Type[] Test2(Type type)
        {
            var genericArguments = type.GetGenericArguments();
            genericArguments[0] = typeof(Example<>).MakeGenericType(genericArguments[0]);
            return genericArguments;
        }
    }

    public readonly struct Example<T> : IComparable<Example<T>>, IEquatable<Example<T>>
        where T : notnull, IComparable<T>, IEquatable<T>
    {
        public readonly ulong Id;
        public readonly T Value;

        public Example(ulong id, T value)
        {
            Id = id;
            Value = value;
        }

        public readonly int CompareTo(Example<T> other)
        {
            if (Id != other.Id)
                return Id < other.Id ? -1 : 1;

            return Comparer<T>.Default.Compare(Value, other.Value);
        }

        public readonly bool Equals(Example<T> other)
        {
            if (Id != other.Id)
                return false;
            if (!EqualityComparer<T>.Default.Equals(Value, other.Value))
                return false;

            return true;
        }

        public override readonly int GetHashCode() => Id.GetHashCode();
        public override readonly bool Equals(object? obj) => obj is Example<T> other && Equals(other);
        public override readonly string ToString() => Value.ToString() + "/" + Id.ToString();

        public static bool operator ==(Example<T> a, Example<T> b) => a.Equals(b);
        public static bool operator !=(Example<T> a, Example<T> b) => !a.Equals(b);
    }
}
```

**Expected IL**:
```msil
    IL_000a:  ldtoken    RoslynBug.Example`1
    IL_000f:  call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
```

**Actual IL**:
```msil
    IL_000a:  ldtoken    valuetype RoslynBug.Example`1<!0>
    IL_000f:  call       class [System.Runtime]System.Type [System.Runtime]System.Type::GetTypeFromHandle(valuetype [System.Runtime]System.RuntimeTypeHandle)
```

**Comment by LeroyK** [comment]
It might be a different issue, because I can actually work around the issue by using ‘’typeof(Example<long>).GetGenericTypeDefinition()’’.

**Comment by cston** [comment]
This is another instance of #38183. The program compiles and runs successfully with the fix in https://github.com/dotnet/roslyn/pull/38249.

**Comment by LeroyK** [comment]
Can we reopen this issue? It is still present in VS2019 16.3 Preview 3 / .NET Core 3 preview9. This example builds fine, but my regular project still compiles incorrect MSIL as described above.

I am trying to create a repro, but it's probably caused by an optimization in the compiler. Perhaps one of the Roslyn devs know how to artificially trigger different optimizations that might cause this?

**Comment by LeroyK** [comment]
Here's a new repro:

```csproj
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <LangVersion>preview</LangVersion>
    <Nullable>enable</Nullable>
    <Deterministic>true</Deterministic>
    <PathMap>$(SolutionDir)=$(SolutionName)</PathMap>
    <Features>pdb-path-determinism</Features>
    <VersionPrefix>0.1.0</VersionPrefix>
    <RuntimeIdentifier>win10-x64</RuntimeIdentifier>
    <PlatformTarget>x64</PlatformTarget>
    <TargetLatestRuntimePatch>true</TargetLatestRuntimePatch>
    <TieredCompilation>true</TieredCompilation>
    <TieredCompilationQuickJit>true</TieredCompilationQuickJit>
    <LinkDuringPublish>false</LinkDuringPublish>
    <CrossGenDuringPublish>false</CrossGenDuringPublish>
    <PublishReadyToRun>false</PublishReadyToRun>
  </PropertyGroup>

</Project>
```

```c#
using System;
using System.Collections.Generic;

namespace RoslynBug
{
    class Program
    {
        static readonly Type s_exampleGenericType = typeof(Example<>);
        static readonly Type s_exampleGenericType2 = typeof(IDictionary<,>);
        static readonly Type s_exampleGenericType3 = typeof(List<>);

        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
            Test(typeof(IDictionary<string, long>));
        }

        static Type[] Test(Type type)
        {
            var genericArguments = type.GetGenericArguments();
            genericArguments[0] = s_exampleGenericType.MakeGenericType(genericArguments[0]);
            return genericArguments;
        }

        static Type[] Test2(Type type)
        {
            var genericArguments = type.GetGenericArguments();
            genericArguments[0] = s_exampleGenericType.MakeGenericType(genericArguments[0]);
            return genericArguments;
        }
    }

    public readonly struct Example<T> : IComparable<Example<T>>, IEquatable<Example<T>>
        where T : notnull, IComparable<T>, IEquatable<T>
    {
        public readonly ulong Id;
        public readonly T Value;

        public Example(ulong id, T value)
        {
            Id = id;
            Value = value;
        }

        public readonly int CompareTo(Example<T> other)
        {
            if (Id != other.Id)
                return Id < other.Id ? -1 : 1;

            return Comparer<T>.Default.Compare(Value, other.Value);
        }

        public readonly bool Equals(Example<T> other)
        {
            if (Id != other.Id)
                return false;
            if (!EqualityComparer<T>.Default.Equals(Value, other.Value))
                return false;

            return true;
        }

        public override readonly int GetHashCode() => Id.GetHashCode();
        public override readonly bool Equals(object? obj) => obj is Example<T> other && Equals(other);
        public override readonly string ToString() => Value.ToString() + "/" + Id.ToString();

        public static bool operator ==(Example<T> a, Example<T> b) => a.Equals(b);
        public static bool operator !=(Example<T> a, Example<T> b) => !a.Equals(b);
    }
}
```

**Comment by gafter** [comment]
I will reopen until we confirm the fix in a public build.

**Comment by gafter** [comment]
> Can we reopen this issue? It is still present in VS2019 16.3 Preview 3 / .NET Core 3 preview9. 

Yes, as you can see the fix https://github.com/dotnet/roslyn/pull/38249 will be available in 16.4 Preview 1.

**Comment by LeroyK** [comment]
So a compiler bug that causes invalid IL and BadImageFormatExceptions (at runtime) is not included in the next GA release (16.3)? Shouldn't this have top priority?

**Comment by cston** [comment]
The fix was subsequently merged into 16.3 (see https://github.com/dotnet/roslyn/pull/38396) but that change did not make it into preview 3. It should be available in the next preview.

I've confirmed the compiler generates valid IL for the latest repro with the fix.

**Comment by cston** [comment]
@LeroyK, thanks for reporting this issue.
